### PR TITLE
feat: add func for git branch

### PR DESCRIPTION
**Pull request by l-qing**
# Changes

* feat: add func for git branch
* chore: generate mock package for plugin

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [`spec` PR link](https://github.com/katanomi/spec) included
- [x] Follows the [commit message standard](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md#commits)
- [x] Meets the [contributing guidelines](https://github.com/katanomi/spec/blob/main/CONTRIBUTING.md) (including
  functionality, content, code)
- [x] Test cases with documentation and functionality works as expected using current and related github repos (MUST deploy and check)
- [x] Release notes block below has been filled in or deleted (only if no user facing changes)

### CODE REVIEW

File: apis/meta/v1alpha1/gitbranch_types.go [hunk 1]
```diff
@@ -77,3 +77,19 @@ func GitBranchResourceAttributes(verb string) authv1.ResourceAttributes {
 		Verb:     verb,
 	}
 }
+
+// IsProtected returns true if the branch is protected
+func (r *GitBranch) IsProtected() bool {
```
Comment (l-qing) [inline apis/meta/v1alpha1/gitbranch_types.go:82]: 这次添加的2个方法。

File: plugin/client/base/client.go [hunk 2]
```diff
@@ -233,3 +234,23 @@ func GetSubResourcesOptionsFromRequest(req *restful.Request) (opts metav1alpha1.
 	}
 	return
 }
+
+// GetGitRepoInfo try to get the project and repository from the git address.
+func GetGitRepoInfo(gitAddress string) (host string, gitRepo metav1alpha1.GitRepo, err error) {
```
Comment (l-qing) [inline plugin/client/base/client.go:239]: 从 v1 client 移动到了 base 中，方便复用。

File: plugin/types/artifact.go [hunk 3]
```diff
@@ -23,6 +23,8 @@ import (
 	metav1alpha1 "github.com/katanomi/pkg/apis/meta/v1alpha1"
 )
 
+//go:generate mockgen -package=types -destination=../../testing/mock/github.com/katanomi/pkg/plugin/types/artifact.go github.com/katanomi/pkg/plugin/types ArtifactLister,ArtifactGetter,ArtifactDeleter,ProjectArtifactLister,ProjectArtifactGetter,ProjectArtifactDeleter,ProjectArtifactUploader,ProjectArtifactFileGetter,ArtifactTagDeleter,ArtifactTriggerRegister
```
Comment (l-qing) [inline plugin/types/artifact.go:26]: 统一添加，按需使用。

File: plugin/types/plugin_services.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package types
+
+//go:generate mockgen -package=types -destination=../../testing/mock/github.com/katanomi/pkg/plugin/types/plugin_services.go github.com/katanomi/pkg/plugin/types PluginServices
+
+// PluginServices contains all the services that a plugin can use
+type PluginServices interface {
```
Comment (l-qing) [inline plugin/types/plugin_services.go:22]: 构建一个大而全的接口，方便直接 mock pluginclient 来写单测。

有些接口没有实现，就把对应接口注释掉了。
如果后期有需要，或者实现了，可以再放开。避免一个个去找还有哪些没实现。

File: plugin/client/base/client.go (hunk 2 above)
Comment (nanjingfm) [reply plugin/client/base/client.go:239]: image

可以放在这里，本身这个方法和 plugin 是没啥关系的

**Review by danielfbm — APPROVED** [review]

**Review by nanjingfm — APPROVED** [review]

**Pull request merged**